// local/entity/query.js
module.exports = Class.create({

    queries: {},

    /**
     * @param name string
     * @param args array  Arguments defined by registering this query. Custom can be added via ´.param(name, key)´
     * @returns {*|{submit, with}}
     */
    query: function(name, args)
    {
        var query = this.queries[name],
            data  = {};

        query.data.forEach(function(name, i) {
            data[name] = args[i];
        });

        return this._createQuery(query.url, data);
    },

    register: function(name, url, data)
    {
        this.queries[name] = {
            url: url,
            data: data
        };
    },

    _createQuery: function(url, data)
    {
        data = data || {};

        return {
            submit: function(callback)
            {
                return $.get(url, data, callback);
            },
            with: function(key)
            {
                if(!data.with)
                {
                    data.with = [];
                }

                data.with.push(key);
                return this;
            },
            only: function(key)
            {
                data.only = key;
                return this;
            },
            param: function(key, value)
            {
                data[key] = value;
                return this;
            },
            params: function(params)
            {
                data = $.extend(data, params);
                return this;
            }
        };
    }
    
});
