// local/form_builder.js
module.exports = Class.create({

    /**
     * Field { label, name- field_type (field_data) }
     * @param field
     */
    create: function(field)
    {
        return [
            this.label(field.label, field.name),
            this[field.field_type.type](field.name, field.field_type.data)
        ];
    },

    select: function(name, properties)
    {
        var $select = $('<select />', {
            name: name
        });

        properties.values.forEach(function(item) {
            $select.append($('<option />', {
                value: item.value,
                html: item.label || item.name
            }))
        });

        return $select;
    },

    checkbox: function(name, properties)
    {
        return $('<input />', {
            type: 'checkbox',
            checked: properties.checked || false,
            name: name
        });
    },

    html: function(name)
    {
        return $('<textarea />', {
            name: name
        });
    },

    email: function(name)
    {
        return $('<input />', {
            type: 'email',
            maxLength: 255,
            name: name
        });
    },

    textarea: function(name, properties)
    {
        return $('<textarea />', {
            maxLength: properties.maxLength || 0,
            name: name
        })
    },

    text: function(name, properties)
    {
        return $('<input />', {
            type: 'text',
            maxLength: properties.maxLength || 0,
            name: name
        });
    },

    number: function(name, properties)
    {
        return $('<input />', {
            type: 'number',
            minValue: properties.min || 0,
            maxValue: properties.max || Number.MAX_VALUE,
            name: name
        });
    },

    label: function(label, fieldName)
    {
        return $('<label />', {
            for: fieldName,
            html: label
        })
    }

});
