// local/table.js
/**
 * Component to build and modify simple DOM tables.
 */
module.exports = require('class-js2').create({

    columns: [],
    rows: [],

    $parent: null,
    $table: null,

    constructor: function()
    {
        var me = this;

        me.columns = [];
        me.$table  = $('<table />');

        me.$table.$head = $('<thead />', {
            html: $('<tr />')
        }).appendTo(me.$table);

        me.$table.$body = $('<tbody />', {

        }).appendTo(me.$table);
    },
    reset: function()
    {
        var me = this;

        me.columns = [];
        me.rows    = [];
        me.$parent = null;

        me.$table.remove();
        me.$table  = null;
    },
    insert: function($parent)
    {
        if(!this.$parent)
        {
            this.$parent = $parent;
            this.$table.appendTo(this.$parent);
        }
    },
    addColumn: function(name, description)
    {
        var $column = $('<th />', {
            html: description
        });

        $column.appendTo(this.$table.$head.find('tr'));
        $column.name = name;

        this.columns.push($column);

        return $column;
    },
    getColumn: function(name)
    {
        for(var i = 0; i < this.columns.length; i++)
        {
            if(this.columns[i].name == name)
            {
                return this.columns[i];
            }
        }
    },
    addRow: function(row)
    {
        var me   = this,
            $row = $('<tr />');

        $row.appendTo(me.$table.$body);
        $row.fields = [];
        me.rows.push($row);

        me.columns.forEach(function($column) {
            if(row[$column.name])
            {
                var $field = $('<td />', {
                    html: row[$column.name]
                });

                $field.appendTo($row);
                $row.fields.push($field);
            }
        });

        return $row;
    },
    addRows: function(rows)
    {
        var me = this;

        rows.forEach(function(row) {
            me.addRow(row);
        });
    },
    clearRows: function()
    {
        var me = this;

        me.rows = [];
        me.$table.$body.empty();
    }
});
