// local/table.js
/**
 * Component to build and modify simple DOM tables.
 */
module.exports = require('classjs').define({
    properties:
    {
        columns: 'Array',

        $parent: 'Object',
        $table: 'Object',

        $tableHead: 'Object',
        $tableBody: 'Object'
    },
    constructor: function()
    {
        this.columns = [];

        this.$table     = $('<table />');

        this.$tableHead = $('<thead />', {
            html: $('<tr />')
        }).appendTo(this.$table);

        this.$tableBody = $('<tbody />', {

        }).appendTo(this.$table);
    },
    members:
    {
        insert: function($parent)
        {
            if(!this.$parent)
            {
                this.$parent = $parent;
                this.$table.appendTo(this.$parent);
            }
        },
        addColumn: function(name, description)
        {
            var $column = $('<th />', {
                html: description
            });

            $column.appendTo(this.$tableHead.find('tr'));
            $column.name = name;

            this.columns.push($column);

            return $column;
        },
        getColumn: function(name)
        {
            for(var i = 0; i < this.columns.length; i++)
            {
                if(this.columns[i].name == name)
                {
                    return this.columns[i];
                }
            }
        }
    }
});
