// local/configuration/form_loader.js
module.exports = Class.create({

    $container: null,

    config: null,

    constructor: function($container, config)
    {
        var me = this;

        me.$container = $container;
        me.config     = config;
    },

    load: function()
    {
        var me  = this;

        me.$container.empty();
        me.$container.append(me.createHeader());
        me.$container.append(me.createFormContainer());
    },

    createFormContainer: function()
    {
        var me = this,
            $container = $('<div />', {
                class: 'config--form'
            });

        Querier.query('config', [ me.config.id ]).only('fields').submit(function(response) {
            response.fields.forEach(function(field) {
                field.field_type = JSON.parse(field.field_type);

                $container.append(FormBuilder.create(field));
            })
        });

        return $container;
    },

    getTitle: function()
    {
        return 'Create a new entity for "' + this.config.name + '"';
    },

    createHeader: function()
    {
        var me         = this,
            $container = $('<div />', {
                'class': 'config--header--container'
            }),
            $header    = $('<div />', {
                'class': 'header--label',
                'html': me.getTitle()
            });

        $container.html($header);

        return $container;
    },

});
