// local/configuration/form_loader.js
module.exports = Class.create({

    $container: null,

    config: null,

    fields: [],

    recordId: null,

    record: null,

    constructor: function($container, config)
    {
        var me = this;

        me.$container = $container;
        me.config     = config;
    },

    load: function(recordId)
    {
        var me  = this;

        me.recordId = recordId;

        if(me.recordId)
        {
            $.get(__url('config/loadRecord'), { id: me.config.id, rowId: recordId }, function(response) {
                me.record = response.record;
                me.fillContainer();
            })
        }
        else
        {
            me.fillContainer();
        }
    },

    fillContainer: function()
    {
        var me = this;

        me.$container.empty();
        me.$container.append(me.createHeader());
        me.$container.append(me.createFormContainer());
    },

    createFormContainer: function()
    {
        var me = this,
            $container = $('<div />', {
                class: 'config--form'
            });

        Querier.query('config', [ me.config.id ]).only('fields').submit(function(response) {
            response.fields.forEach(function(field) {
                field.field_type = JSON.parse(field.field_type);

                var $field = FormBuilder.create(field);
                me.fields.push($field[1]);

                if(me.record && me.record[field.name])
                {
                    $field[1].val(me.record[field.name]);
                }

                $container.append($field);
            })
        });

        return $container;
    },

    getTitle: function()
    {
        return this.recordId ?
            'Edit an entity for "' + this.config.label + '"' :
            'Create a new entity for "' + this.config.label + '"';
    },

    createHeader: function()
    {
        var me         = this,
            $container = $('<div />', {
                'class': 'config--header--container'
            }),
            $header    = $('<div />', {
                'class': 'header--label',
                'html': me.getTitle()
            }),
            $button    = $('<button />', {
                'class': 'header--button',
                'html': 'Save'
            });

        $button.on('click', function() {
            var data = {
                id: me.config.id,
                data: {}
            };

            me.fields.forEach(function($field) {
                data.data[$field.prop('name')] = $field.val();
            });

            if(me.recordId)
            {
                data.rowId = me.recordId;
            }

            $.get(__url('config/save'), data, function(response) {
                (new (require('local/configuration/loader'))(me.config.id)).load();
            })
        });

        $container.html([ $header, $button ]);

        return $container;
    },

});
