// local/configuration/form_loader.js
module.exports = Class.create({

    $container: null,

    config: null,

    fields: [],

    constructor: function($container, config)
    {
        var me = this;

        me.$container = $container;
        me.config     = config;
    },

    load: function()
    {
        var me  = this;

        me.$container.empty();
        me.$container.append(me.createHeader());
        me.$container.append(me.createFormContainer());
    },

    createFormContainer: function()
    {
        var me = this,
            $container = $('<div />', {
                class: 'config--form'
            });

        Querier.query('config', [ me.config.id ]).only('fields').submit(function(response) {
            response.fields.forEach(function(field) {
                field.field_type = JSON.parse(field.field_type);

                var $field = FormBuilder.create(field);
                me.fields.push($field[1]);

                $container.append($field);
            })
        });

        return $container;
    },

    getTitle: function()
    {
        return 'Create a new entity for "' + this.config.label + '"';
    },

    createHeader: function()
    {
        var me         = this,
            $container = $('<div />', {
                'class': 'config--header--container'
            }),
            $header    = $('<div />', {
                'class': 'header--label',
                'html': me.getTitle()
            }),
            $button    = $('<button />', {
                'class': 'header--button',
                'html': 'Save'
            });

        $button.on('click', function() {
            me.fields.forEach(function($field) {
                console.log($field.prop('name'), $field.val());
            });
        });

        $container.html([ $header, $button ]);

        return $container;
    },

});
