// local/configuration/loader.js
module.exports = require('class-js2').create({

    configId: -1,

    params:
    {
        orderBy: 'id',
        order: 'DESC',
        filterBy: '',
        limit: 15,
        offset: 0
    },

    $container: null,
    table: null,

    url:
    {
        listing: '',
        listingColumns: ''
    },

    constructor: function(configId)
    {
        var me = this;

        me.configId   = configId;
        me.$container = $('.configuration--container');

        me.url.listing        = me.$container.attr('data-listingUrl');
        me.url.listingColumns = me.$container.attr('data-listingColumnsUrl');
    },
    load: function()
    {
        var me = this;

        me.$container.empty();

        me.createTable(me.loadData.bind(me));
    },

    loadData: function()
    {
        var me = this;

        $.get(me.url.listing, {
            id: me.configId,
            orderBy: me.params.orderBy,
            order: me.params.order,
            filterBy: me.params.filterBy,
            limit: me.params.limit,
            offset: me.params.offset
        }, function(response) {
            me.table.addRows(response.data);
        });
    },

    createTable: function(done)
    {
        var me = this;

        me.table = new (require('local/table'));
        me.table.insert(me.$container);

        me.loadColumns(function(columns) {
            columns.forEach(function(column) {
                me.table.addColumn(column.name, column.label);
            });

            me.table.addColumn('actions', 'Actions');

            done();
        });
    },

    loadColumns: function(done)
    {
        var me = this;

        $.get(me.url.listingColumns, {id: me.configId}, function(response) {
            if(response.success)
            {
                done.apply(me, [response.data]);
            }
        });
    }
});
